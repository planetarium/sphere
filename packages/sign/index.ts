import { toChecksum } from "./util";
import { encode, decode, BencodexDict } from "bencodex";
import { keccak_256 } from "@noble/hashes/sha3";
import { bytesToHex, hexToBytes } from "@noble/hashes/utils";
import { Buffer } from "buffer/";

/**
 * Account is a implementation to perform cryptographic signing and verifying them.
 *
 * Accounts can be provided from the @planetarium/account-* packages or third-parties
 * using the name `planetarium-account-*`.
 */
export interface Account {
  /**
   * Account interface is versioned to safely make changes to the account interface.
   * The users should make sure if the account is compatible with the sign interface
   * they are using.
   */
  readonly VERSION: number;

  getPublicKey(): Promise<ArrayBuffer>;
  /**
   * Sign a given bytes. The function must return a valid ECDSA signature.
   *
   * @param data A payload to sign.
   */
  sign(hash: ArrayBuffer): Promise<ArrayBuffer>;
}

export const ACCOUNT_VERSION = 0;

export async function signTransaction(
  tx: string,
  account: Account
): Promise<string> {
  if (account.VERSION !== ACCOUNT_VERSION)
    throw new Error("The Account interface version doesn't match.");

  const txBinary = hexToBytes(tx);
  const decodedTx: BencodexDict = decode(new Buffer(txBinary));

  const hash = await crypto.subtle.digest("SHA-256", txBinary);
  const signature = await account.sign(hash);

  decodedTx.set(new Buffer([0x53]), signature);

  return encode(decodedTx)?.toString("hex");
}

export async function deriveAddress(account: Account) {
  const publicKey = await account.getPublicKey();
  return (
    "0x" +
    toChecksum(
      bytesToHex(keccak_256(new Uint8Array(publicKey).slice(1))).slice(-40)
    )
  );
}
