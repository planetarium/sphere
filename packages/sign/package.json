{
  "name": "@planetarium/sign",
  "version": "0.0.0",
  "license": "LGPL",
  "source": "index.ts",
  "module": "dist/index.mjs",
  "main": "dist/index.cjs",
  "types": "dist/index.d.ts",
  "dependencies": {
    "@noble/hashes": "^1.1.2",
    "bencodex": "^0.1.1",
    "buffer": "^6.0.3"
  },
  "devDependencies": {
    "nanobundle": "^0.0.28"
  },
  "scripts": {
    "build": "nanobundle build"
  }
}
