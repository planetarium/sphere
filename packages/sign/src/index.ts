import { toChecksum } from "./util";
import { encode, decode, BencodexDict } from "bencodex";
import { keccak_256 } from "@noble/hashes/sha3";
import { bytesToHex, hexToBytes } from "@noble/hashes/utils";
import { Buffer } from "#buffer";
import * as crypto from "#webcrypto";

/**
 * Account is a implementation to perform cryptographic signing and verifying them.
 *
 * Accounts can be provided from the @planetarium/account-* packages or third-parties
 * using the name `planetarium-account-*`.
 */
export interface Account {
  /**
   * Account interface is versioned to safely make changes to the account interface.
   * The users should make sure if the account is compatible with the sign interface
   * they are using.
   */
  readonly VERSION: number;

  getPublicKey(isCompressed?: boolean): Promise<Uint8Array>;
  /**
   * Sign a given bytes. The function must return a valid ECDSA signature using
   * 'SECP256K1' curve.
   * 
   * @param data A payload to sign.
   */
  sign(hash: Uint8Array): Promise<Uint8Array>;
}

export const ACCOUNT_VERSION = 0;

export async function getEncodedPublicKey(
  account: Account
): Promise<string> {
  return Buffer.from(await account.getPublicKey(true)).toString("hex");
}

export async function signTransaction(
  tx: string,
  account: Account
): Promise<string> {
  if (account.VERSION !== ACCOUNT_VERSION)
    throw new Error("The Account interface version doesn't match.");

  const txBinary = hexToBytes(tx);
  const decodedTx: BencodexDict = decode(new Buffer(txBinary));

  const hash = await crypto.subtle.digest("SHA-256", txBinary);
  const signature = await account.sign(new Uint8Array(hash));

  if (
    Array.from(decodedTx.entries()).some(
      ([key]: [Buffer, unknown]) => key[0] === 0x53
    )
  )
    throw new Error("Already signed.");

  decodedTx.set(new Uint8Array([0x53]), signature.buffer);

  return encode(decodedTx)?.toString("hex");
}

export async function deriveAddress(account: Account) {
  const publicKey = await account.getPublicKey();
  return (
    "0x" + toChecksum(bytesToHex(keccak_256(publicKey.slice(1))).slice(-40))
  );
}
