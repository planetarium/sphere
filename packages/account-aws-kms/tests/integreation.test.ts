import { expect, test, beforeAll, describe } from "vitest";
import { KMSClient } from "@aws-sdk/client-kms";
import { fromIni } from "@aws-sdk/credential-providers";
import { deriveAddress, signTransaction } from "@planetarium/sign";
import { createAccount } from "..";

if (globalThis.crypto == null) {
  // FIXME: This is a workaround for the lack of Web Crypto API in Vitest.
  // @ts-ignore
  globalThis.crypto = require("node:crypto");
}

const KEY_ID =
  import.meta.env.AWS_KMS_KEY_ID || "add1894e-86b8-4ffb-aa76-2186b3adc598";

const kmsClient = new KMSClient({
  credentials: fromIni({ profile: import.meta.env.AWS_PROFILE }),
});

test("create a account from keyId", () => {
  expect(() => createAccount(kmsClient, KEY_ID)).to.not.throw;
});

describe("account interface", () => {
  const account = createAccount(kmsClient, KEY_ID);

  describe("details", () => {
    test("publicKey", async () => expect(await account.getPublicKey()).toMatchInlineSnapshot(`
      Uint8Array [
        4,
        170,
        23,
        202,
        117,
        194,
        99,
        225,
        118,
        0,
        62,
        20,
        240,
        251,
        142,
        112,
        145,
        223,
        182,
        252,
        115,
        205,
        91,
        215,
        188,
        12,
        12,
        228,
        2,
        42,
        107,
        33,
        164,
        236,
        27,
        123,
        246,
        228,
        10,
        33,
        13,
        209,
        226,
        208,
        93,
        124,
        134,
        252,
        31,
        238,
        62,
        169,
        149,
        179,
        171,
        99,
        86,
        25,
        225,
        129,
        112,
        171,
        7,
        97,
        213,
      ]
    `))
  })

  test("deriveAddress", async () =>
    expect(await deriveAddress(account)).toMatchInlineSnapshot(
      '"0xA0208696A8E001fEb28b4EF94B9B0029B9C31c06"'
    ));
  test("signTransaction", async () => {
    expect(
      await signTransaction(
        "64313a616c6475373a747970655f696475363a7374616b653275363a76616c7565736475323a616d69316575323a696431363a64a268c8248da04b954c9141395e52f8656565313a6733323a4582250d0da33b06779a8475d283d5dd210c683b9b999d74d03fac4f58fa6bce313a6e693065313a7036353a04aa17ca75c263e176003e14f0fb8e7091dfb6fc73cd5bd7bc0c0ce4022a6b21a4ec1b7bf6e40a210dd1e2d05d7c86fc1fee3ea995b3ab635619e18170ab0761d5313a7332303aa0208696a8e001feb28b4ef94b9b0029b9c31c06313a747532373a323032322d31302d30355430353a33383a31342e3630363836365a313a756c32303a037afa4bec223b5f6f2f41f6a475f2d1dff1ad7332303aa0208696a8e001feb28b4ef94b9b0029b9c31c066565",
        account
      )
    ).toMatchSnapshot();
  });
});
