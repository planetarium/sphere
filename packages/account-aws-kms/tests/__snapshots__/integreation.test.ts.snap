// Vitest Snapshot v1

exports[`account interface > signTransaction 1`] = `"64313a536475313a306934386575313a316937306575323a3130693132386575323a313169356575323a3132693134366575323a31336938316575323a31346933326575323a31356932326575323a3136693134306575323a31376938376575323a31386939346575323a31396936316575313a3269326575323a3230693130366575323a32316932306575323a3232693132366575323a3233693131356575323a32346933376575323a32356934306575323a3236693131396575323a3237693135336575323a3238693139336575323a3239693132366575313a336933336575323a33306937346575323a33316936316575323a33326934306575323a3333693139396575323a3334693133386575323a3335693135346575323a33366936316575323a333769326575323a33386933336575323a333969306575313a3469306575323a3430693139366575323a343169386575323a34326939346575323a3433693136376575323a3434693133356575323a3435693230356575323a34366935336575323a34376931306575323a3438693130366575323a3439693232396575313a35693135306575323a35306935356575323a3531693130336575323a35326936386575323a3533693136326575323a3534693138376575323a3535693230306575323a35366934356575323a35376938316575323a35386932366575323a3539693234336575313a3669316575323a3630693233366575323a3631693135306575323a3632693130356575323a3633693133356575323a36346931376575323a36356932326575323a36366932346575323a3637693235356575323a3638693137366575323a3639693136376575313a37693138366575323a3730693135376575323a37316937386575313a38693233306575313a39693133356565313a616c6475373a747970655f696475363a7374616b653275363a76616c7565736475323a616d69316575323a696431363a64a268c8248da04b954c9141395e52f8656565313a6733323a4582250d0da33b06779a8475d283d5dd210c683b9b999d74d03fac4f58fa6bce313a6e693065313a7036353a04aa17ca75c263e176003e14f0fb8e7091dfb6fc73cd5bd7bc0c0ce4022a6b21a4ec1b7bf6e40a210dd1e2d05d7c86fc1fee3ea995b3ab635619e18170ab0761d5313a7332303aa0208696a8e001feb28b4ef94b9b0029b9c31c06313a747532373a323032322d31302d30355430353a33383a31342e3630363836365a313a756c32303a037afa4bec223b5f6f2f41f6a475f2d1dff1ad7332303aa0208696a8e001feb28b4ef94b9b0029b9c31c066565"`;
